### PR TITLE
Faster wires

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
New, faster wires, and a few touches here and there.

**Description of the Change:**
Wires now works with sets and it's unordered in its internal operation. When it has to return modes in an order it will choose increasing order. 

**Benefits:**
Much simpler implementation, super fast

### CODE REVIEW

File: mrmustard/lab_dev/states/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:82]:
```suggestion
    def __init__(self, name: str = "Ket", modes: tuple[int,...] = ()):
```

File: mrmustard/lab_dev/states/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:49]: I think you need to modify Unitary and Channel in the same way

File: mrmustard/lab_dev/states/states.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:118]: DM and Ket have `modes = tuple[int, ...]`, wires has sets, here there are sequences. Choose one?

File: mrmustard/lab_dev/wires.py
Comment (SamFerracin) [inline mrmustard/lab_dev/wires.py:27]:
```suggestion
    In MrMustard, instances of ``CircuitComponent`` have a ``Wires`` attribute.
```

File: mrmustard/lab_dev/states/states.py
Comment (ziofil) [reply mrmustard/lab_dev/states/states.py:118]: Let's use tuples throughout (small, immutable). Only Wires will have sets. I'm not going to chase after all the type annotations in this PR though.

**Review by SamFerracin — COMMENTED** [review]
Looks very good! 
I left a few minor comments. Other than that, there are a few things to discuss, namely:
- Allowed types for `modes` for the various classes
- What to do with the parameter set (which we may want to "fix" in another PR though, seeing that it's an issue that exists previous to this PR?)
Good to merge after that discussion :)

File: mrmustard/lab_dev/wires.py
Comment (SamFerracin) [inline mrmustard/lab_dev/wires.py:269]: https://github.com/XanaduAI/MrMustard/pull/361#discussion_r1518396850

Comment (ziofil) [reply mrmustard/lab_dev/wires.py:338]: not sure I follow 😅

File: tests/test_lab_dev/test_wires.py
Comment (SamFerracin) [inline tests/test_lab_dev/test_wires.py:98]: @ziofil Could you write a test to show how one can set the ids?

- Why it's useful to set the `id`s:
To connect 2 components while constructing the graph, which is useful to either validate the user-provided path (simulator) or to find the optimal path

- How we do this now:
By manually changing the ids of the wires that contract

File: mrmustard/lab_dev/circuit_components.py
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components.py:79]: :+1:  pretty cool
Just a reminder that we will need to write the same piece of code to reorder the parameter set

File: mrmustard/lab_dev/states/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:65]: I don't know why we're not using `_from_attributes` here (I probably forgot to update this), but I believe that `dm = DM._from_attributes("", component.representation, component.wires)` should do the job of these 4 lines

File: mrmustard/lab_dev/states/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:108]: ``dm = DM._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:65]: ``unitary = Unitary._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:71]: ``channel = Channel._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/transformations/base.py
Comment (SamFerracin) [inline mrmustard/lab_dev/transformations/base.py:114]: ``channel = Channel._from_attributes("", component.representation, component.wires)``

File: mrmustard/lab_dev/states/states.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:36]:
```suggestion
    If ``x`` and/or ``y`` are ``Sequence``\s, their length must be equal to `1` or `N`. If their length is equal to `1`,
```

File: mrmustard/lab_dev/states/states.py
Comment (SamFerracin) [inline mrmustard/lab_dev/states/states.py:168]:
```suggestion
        modes: Sequence[int],
```

**Review by SamFerracin — COMMENTED** [review]
Great job!
Most of the comments are about docs, or things that I forgot in previous PRs :sweat_smile: 

The biggest question I have is related to the `id`s: how can we set them, when we want to connect 2 components? See the comment in `test_wires.py`

Can you also pull develop?

File: mrmustard/lab_dev/wires.py
Comment (ziofil) [reply mrmustard/lab_dev/wires.py:201]: nope

**Comment by SamFerracin** [comment]
I fixed the tests and addressed the last comments.
However, I wasn't able to address the most important one, namely how to change the ids ([this comment](https://github.com/XanaduAI/MrMustard/pull/362#discussion_r1528624971)). I am worried that they may not be changeable because of the caching. Let's make sure before we merge

**Review by SamFerracin — APPROVED** [review]
:rocket:

**Pull request merged**
